compiler/optimizations/forallOptimizations.cpp

```diff
@@ -1945,14 +1981,39 @@ static CallExpr *findMaybeAggAssignInBlock(BlockStmt *block) {
 
 // called during resolution when we revert an aggregation
 static void removeAggregatorFromFunction(Symbol *aggregator, FnSymbol *parent) {
-  // remove the definition
-  aggregator->defPoint->remove();
 
-  // find and remove other SymExprs within the function
+  // find other SymExprs within the function and remove the aggregator if it is
+  // not being used by other primitives, or its `copy` function.
+  
+  std::vector<SymExpr *> symExprsToCheck;
+  collectSymExprsFor(parent, aggregator, symExprsToCheck);
+
   std::vector<SymExpr *> symExprsToRemove;
-  collectSymExprsFor(parent, aggregator, symExprsToRemove);
-  for_vector(SymExpr, se, symExprsToRemove) {
-    se->remove();
+  bool shouldRemove = true;
+
+  for_vector(SymExpr, se, symExprsToCheck) {
+    if (CallExpr *parentCall = toCallExpr(se->parentExpr)) {
+      if (parentCall->isPrimitive(PRIM_MAYBE_AGGREGATE_ASSIGN)) {
+        shouldRemove = false;
+        break;
+      }
+
+      if (parentCall->isNamed("copy")) {
```

Comment:
This seems like a really weird case. Could you say more about what's going on here?

It's not great to compare names in the compiler but in this case it seems worse because the name "copy" could be defined by a user. So I am worried this will not be correct in that event.

Reply:
Understandable.

First, the symbol we check here is the compiler-generated aggregator. And `copy` is a method defined on the aggregator. This basically checks whether we have written any `copy(aggregator, src, dst);`. The aggregator symbol is not accessible to the user, so I don't think they could write anything that can throw this check off.

Second, there can be ways we can reduce this ugliness:
- rename `copy` to `chpl_copy` (still a bit ugly)
- add a pragma to the aggregator's copy function, and check that flag here.
- something else?

What do you think?

Reply:
Even simpler - If the `copy` call is supposed to be a method on the aggregator maybe you could check here that the CallExpr is a method call and the Aggregator is the receiver? I don't see any of that here. (If it's an assumption this code makes, you could add some INT_ASSERTs).